### DISCUSSION TITLE
Cannot find main class

### DISCUSSION THREAD

**Discussion by PerryCameron** [wiverson/maven-jpackage-template#81] [General]
So I used your template to update an older java1.8 JavaFX application. 

With Eclipse run configurations work
Also 'mvn javafx:run' works
Also 'mvn clean install' works

Here is the issue:
After dropping the app into applications and then double clicking on the app it fails to launch. I see the icon appear but then quickly disappear. The first thing I did was check to see if there was a log in my home directory. There was not so that told me the app wasn't getting that far. I then launched it from the terminal /Applications/Halyard.app/Contents/MacOS/Halyard

What I got was: Missing JavaFX application class com.ecsail.BaseApplication

BTW: /Applications/TestApp.app/Contents/MacOS/TestApp launches TestApp.app with no issues.

So then I opened 2 instances of finder and compared Halyard.app with TestApp.app, by right clicking and "show contents". 
The jars are located under Contents/app/

Contents/app/
automation-1.11-8.jar
commons-lang3-3.5.jar
generex-1.0.2.jar
javafaker-1.0.2.jar 
maven-jpackage-template-1.0-SNAPSHOT jar. 

The maven-jpackage-template jar is the code in our apps.

Anyway, the class BaseApplication was present in the correct location in both of our apps. In fact I compared everything in our two jars and everything looked perfect in both, so this is a head scratcher.

Also, there is a config file in that directory...

**Here is the one for TestApp**
 [Application]
app.classpath=$APPDIR/maven-jpackage-template-1.0-SNAPSHOT.jar
app.mainclass=com.changenode.BaseApplication
app.classpath=$APPDIR/automaton-1.11-8.jar
app.classpath=$APPDIR/commons-lang3-3.5.jar
app.classpath=$APPDIR/generex-1.0.2.jar
app.classpath=$APPDIR/javafaker-1.0.2.jar
app.classpath=$APPDIR/snakeyaml-1.23-android.jar

[JavaOptions]
java-options=-Djpackage.app-version=1.0.0

**Here is the one for Halyard[Application]**
app.classpath=$APPDIR/maven-jpackage-template-1.0-SNAPSHOT.jar
app.mainclass=com.ecsail.BaseApplication
app.classpath=$APPDIR/automaton-1.11-8.jar
app.classpath=$APPDIR/commons-lang3-3.5.jar
app.classpath=$APPDIR/generex-1.0.2.jar
app.classpath=$APPDIR/javafaker-1.0.2.jar
app.classpath=$APPDIR/jsch-0.1.55.jar
app.classpath=$APPDIR/mysql-connector-java-8.0.30.jar
app.classpath=$APPDIR/protobuf-java-3.19.4.jar
app.classpath=$APPDIR/slf4j-api-2.0.1.jar
app.classpath=$APPDIR/slf4j-simple-2.0.1.jar
app.classpath=$APPDIR/snakeyaml-1.23-android.jar

[JavaOptions]
java-options=-Djpackage.app-version=1.0.0

So the app runs fine in run configurations, mvn javafx:run, and the jar compiles correctly to an executable package. Everything looks fine when you drill into the package. Have any ideas on this one?

BTW I love your template, and am learning a ton of stuff here. Thank you for making this.

**Comment by wiverson** [comment]
Hmm. For starters, I would use JDK 17 or 18 with a bundled JavaFX (Liberica or Zulu). Beyond that I'm not too sure, perhaps try unzipping the jars to make sure things are as you expect...

**Reply by PerryCameron** [reply]
According to your README, It read as 18 was required unless using swing. I would have preferred 17 if I knew that would work. Using sdkman, 18 with fx is no longer available, so I went with 19. I learned that 'sdk install java <tab>' would show available versions.

11.0.16-amzn         17.0.4-amzn          17.0.4.1.fx-zulu     19.fx-librca         22.2.r11-nik
11.0.16.1-librca      17.0.4-oracle        18.0.2-sem           19.fx-zulu           22.2.r17-grl
11.0.16.1-ms          17.0.4.1-librca      19-amzn              20.ea.16-open        22.2.r17-nik
11.0.16.1-sapmchn    17.0.4.1-ms          19-librca            20.ea.17-open        8.0.342-amzn
11.0.16.1-sem        17.0.4.1-sapmchn     19-open              22.1.0.1.r11-gln     8.0.345-librca
11.0.16.1-tem        17.0.4.1-sem         19-oracle            22.1.0.1.r17-gln     8.0.345-zulu
11.0.16.1-zulu       17.0.4.1-tem         19-sapmchn           22.1.0.r11-grl       8.0.345.fx-librca
11.0.16.1.fx-librca  17.0.4.1-zulu        19-tem               22.1.0.r17-grl       8.0.345.fx-zulu
11.0.16.1.fx-zulu    17.0.4.1.fx-librca   19-zulu              22.2.r11-grl

Anyway, I created another parallel project and am slowly building it up. I have learned quite a few interesting things. For example 'mvn javafx:run' is not nearly as picky as a packaged result. I had some unused methods with SQLExceptions that javafx:run didn't care about but the packaged result would leave a nasty gram in my home folder, after an abrupt crash.

I also wanted to complement you on one thing. I am not sure if you intended it or not, but when I do a clean install, I love how it builds the package, mounts if for a few seconds (long enough to drag the package over) and then auto ejects. That is so fantastic for quickly testing a new build.

**Reply by wiverson** [reply]
>I love how it builds the package, mounts if for a few seconds (long enough to drag the package over) and then auto ejects.

Ack, don't do that! That's part of jpackage running - jpackage generates the .dmg by mounting it, copying the files over, and then unmounting it. So, if you are grabbing the app while this is running you are likely grabbing an incomplete .app, which would cause all sort of weird errors.

If you want to write a script to do something like copy the contents of the .dmg file I would suggest writing a script and use the [hdiutil](https://ss64.com/osx/hdiutil.html) command to mount the .dmg file. Basically you would write a little shell script to run the mvn build, then use hdiutil to mount the generated .dmg, and then do something (e.g. copy the .app into your /Applications folder).

**Comment by PerryCameron** [comment]
Hi there Will,

So here it is a couple of years later and I have a repo doing the same thing:

**Perry$** /Applications/Halyard.app/Contents/MacOS/Halyard 
Jun 09, 2024 1:25:11 PM com.sun.glass.ui.mac.MacApplication lambda$waitForReactivation$6
WARNING: Timeout while waiting for app reactivation
Missing JavaFX application class org.ecsail.BaseApplication

After a few hours of troubleshooting I found that I could launch the app using the system Java runtime, which happened to be the same one I was using to embed. openjdk version "17.0.4.1" 2022-08-12 LTS

Here is the command I used for that:
**Perry$** java --module-path ../runtime/lib --add-modules javafx.controls,javafx.fxml -cp ../app/maven-jpackage-template-1.1-SNAPSHOT.jar:../app/* org.ecsail.BaseApplication

The application works fine like that, and now I know the problem has something to do with the embedded runtime.

So I began to compare a working package to the non-working package. The interesting thing is that I could find no java binary anywhere in the runtime dir for both the working and non-working package.

My question to you is, without a Java binary, How does this thing work?

**Comment by wiverson** [comment]
Well, if you are on macOS nowadays I would recommend using [SDKMan](https://sdkman.io/jdks) to install a JDK. That will typically install the JDK in a well known/OS standard directory and also add it to the path.

Beyond that, hard to say as depending on a) your OS configuration b) which JDK variant[s] you have installed and c) what your app is trying to do/dependencies it's hard to say.

The idea of this template is that it takes the binaries from the installed JDK and then generates an installer that bundles those binaries. From the perspective of the end user you just get an exe or msi or dmg or whatever and it just looks like an ordinary app. No extra JDK installation needed, it's all bundled inside the app. If you grab the macOS dmg demo app in releases and right click to view the .app guts you'll see the JDK bits inside. 

I'm assuming that your app is https://github.com/PerryCameron/Halyard - looks like you already have a nice clean Maven build. The way that others have gotten this working is by cloning off the assets of the template, then grabbing your existing pom.xml and src and merging them into the template. You can try doing it the other way, of course - incrementally updating your project by grabbing bits and pieces from the template. That's probably going to generate a lot more errors as you go but you'll learn a lot fixing them.

**Reply by PerryCameron** [reply]
I use sdkman. I have figured out that I must have the default system sdk match whatever project I am using. I would think that when I set the SDK in intellij that would be all I have to do, but unfortunately that is not how it works. I suppose I could put an sdk check in the build script, and if it doesn't match have it switch sdk's. 

That version of Halyard you linked there, was the original before I forked it to use your project to package up my application. I have been using your system for a couple of years and it works great. This being said, it is time that I spend a good amount of time to learn Gradle, and make my own custom build scripts. TY for all your hard work Will.

**Reply by EasyG0ing1** [reply]
@PerryCameron - Just a side interest here, if you figure out how to get IntelliJ to switch a project's SDK based on the current default SDK as set using SDKMan, I would be interested in knowing how you did it.

Currently, when I'm using some SDK (These days, I like GraalVM because you can make native-image binaries from your projects so that they don't need a Java VM to execute as they execute as native code to the OS - makes programs much faster and easier to distribute) and when I open a project in IntelliJ, I just hit `CMD+;` and manually change the project/module SDK and keep going. Having that automated would be really nice.

Thanks,

Mike

**Reply by PerryCameron** [reply]
The only way I know how to that is with Gradle, not intellij.  I am not sure that would work for you. Oh wow on the GraalVM, I am going to have to look into that when I get some time.

**Reply by EasyG0ing1** [reply]
@PerryCameron 

> The only way I know how to that is with Gradle, not intellij. I am not sure that would work for you. Oh wow on the GraalVM, I am going to have to look into that when I get some time.

GraalVM will take some time to figure out. It's easy to do when your application is a CLI app, much harder to use when you're trying to compile a GUI based app. But being able to run and distribute Java without the end user needing a JVM at all is a dream come true.
